### PR TITLE
url: reject ASCII control characters and space in host names

### PR DESCRIPTION
**Pull request by bagder**
Host names like "127.0.0.1 moo" would otherwise be accepted by some
getaddrinfo() implementations.

Fixes #2073

### CODE REVIEW

**Comment by bagder** [comment]
This is an alternative take to #2073, and as a bonus it also checks proxy host names etc. No test in here, but @mkauf's test from #2073 should be able to verify this approach as well.

File: lib/url.c [hunk 1]
```diff
@@ -4179,13 +4189,21 @@ static CURLcode create_conn(struct Curl_easy *data,
...
-  if(conn->bits.httpproxy)
-    fix_hostname(conn, &conn->http_proxy.host);
-  if(conn->bits.socksproxy)
-    fix_hostname(conn, &conn->socks_proxy.host);
+  result = fix_hostname(conn, &conn->host);
+  if(result)
+    goto out;
+  if(conn->bits.conn_to_host) {
+    result = fix_hostname(conn, &conn->conn_to_host);
+    goto out;
```
Comment (mkauf) [inline lib/url.c]:
```
goto out;
```
... must be changed to:

```
if(result)
  goto out;
```

File: lib/url.c [hunk 2]
```diff
@@ -4179,13 +4189,21 @@ static CURLcode create_conn(struct Curl_easy *data,
...
+  }
+  if(conn->bits.httpproxy) {
+    result = fix_hostname(conn, &conn->http_proxy.host);
+    goto out;
```
Comment (mkauf) [inline lib/url.c]: same problem here

File: lib/url.c [hunk 3]
```diff
@@ -4179,13 +4189,21 @@ static CURLcode create_conn(struct Curl_easy *data,
...
+  }
+  if(conn->bits.socksproxy) {
+    result = fix_hostname(conn, &conn->socks_proxy.host);
+    goto out;
```
Comment (mkauf) [inline lib/url.c]: same problem here

Comment (bagder) [reply lib/url.c]: I clearly was a bit too eager to get something to show...

**Comment by bagder** [comment]
> it could also report failed IDN conversions. What do you think?

Oh yes! I'll add that too when I fix my initial mistakes.

**Comment by bagder** [comment]
Hah, the new IDN error changes the error for a few test cases so I need to clean those up...

**Comment by bagder** [comment]
Test 1264 was now cherry-picked into this branch from #2073

**Pull request closed without merging**